bundles-lib/src/main/java/org/apache/metron/bundles/BundleClassLoaders.java

```diff
@@ -0,0 +1,353 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.metron.bundles;
+
+import java.net.URISyntaxException;
+import java.util.*;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.vfs2.FileObject;
+import org.apache.commons.vfs2.FileSystemException;
+import org.apache.commons.vfs2.FileSystemManager;
+import org.apache.metron.bundles.bundle.Bundle;
+import org.apache.metron.bundles.bundle.BundleCoordinate;
+import org.apache.metron.bundles.bundle.BundleDetails;
+import org.apache.metron.bundles.util.BundleProperties;
+import org.apache.metron.bundles.util.BundleSelector;
+import org.apache.metron.bundles.util.FileUtils;
+import org.apache.metron.bundles.util.BundleUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A singleton class used to initialize the extension and framework classloaders.
+ */
+public final class BundleClassLoaders {
+
+  private static volatile BundleClassLoaders bundleClassLoaders;
+  private volatile InitContext initContext;
+  private static final Logger logger = LoggerFactory.getLogger(BundleClassLoaders.class);
+
+  private final static class InitContext {
+
+    private final List<FileObject> extensionDirs;
+    private final Map<String, Bundle> bundles;
+
+    private InitContext(
+        final List<FileObject> extensionDirs,
+        final Map<String, Bundle> bundles) {
+      this.extensionDirs = extensionDirs;
+      this.bundles = bundles;
+    }
+  }
+
+  private BundleClassLoaders() {
+  }
+
+  /**
+   * @return The singleton instance of the BundleClassLoaders
+   */
+  public static BundleClassLoaders getInstance() {
+    BundleClassLoaders result = bundleClassLoaders;
+    if (result == null) {
+      synchronized (BundleClassLoaders.class) {
+        result = bundleClassLoaders;
+        if (result == null) {
+          bundleClassLoaders = result = new BundleClassLoaders();
+        }
+      }
+    }
+    return result;
+  }
+
+  public static void reset() {
+    getInstance().unInit();
+  }
+
+  private void unInit() {
+    synchronized (this) {
+      initContext = null;
+    }
+  }
+
+  /**
+   * Initializes and loads the BundleClassLoaders. This method must be called before the rest of the
+   * methods to access the classloaders are called and it can be safely called any number of times
+   * provided the same framework and extension working dirs are used.
```

Comment:
Yes.  The functionality is incomplete at the moment, but that is because we don't have a host that hosts multiple plugins yet.  When you add a parser extension and start it ( METRON-942) you are starting a new topology.  

A complete and formal pattern or implementation around Bundle System ReInit() would be a next step, but the primitives are there.